library/x509_crt.c

```diff
@@ -823,6 +966,13 @@ static int x509_get_crt_ext( unsigned char **p,
                 return( ret );
             break;
 
+        case MBEDTLS_OID_X509_EXT_CERTIFICATE_POLICIES:
+            /* Parse certificate policies type */
+            if( ( ret = x509_get_certificate_policies( p, end_ext_octet,
```

Comment:
Blocker: If this extension is critical and contains at least one policy that is not `anyPolicy`, then we accept it even if we shouldn't because of these policies that we parse but don't interpret/enforce.

Reply:
ok. I was looking for when we return the "Not supported" error, which we don't. This is why i didn't add this check.
But actually, we don't really enforce the policies, and  since we return all the policies to the user, including what we don't interpret in our information function, shouldn't it be the user's responsibility parse \ check additional policies?

Reply:
Yes, it is the user's responsibility, but the user does it with the help of the verification callback. Therefore our responsibility is to fail if we encounter a critical policy that we cannot interpret/enforce. If the user can interpret it, they use the verification callback to do so and to enforce it at the same time.

Reply:
I don't think so, especially with hanno's PR.
If we fail because we can't enforce the policy, then we return from the extension parsing immediately, not continuing to parse other extensions.
I don't htink the verification callback ansers this.
What I meant was, that we return the raw data of the certificate policies, and then the user can parse \ interpret the policy after handshake \ certificate parsing finishes

Reply:
We have two use cases:
1. The user does not need certificate policies that Mbed TLS can't interpret/enforce
2. The user does need certificate policies that Mbed TLS can't interpret/enforce

Case 1. is the default and the most common case. We have a compile time option to support case 2: `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION`. If this option is on, then the parsing does not fail and the user can parse the certificate again in the callback and interpret/enforce the policies in question.

Currently the code always allows unsupported critical certificate policies extension whether  `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION` is on or not. This means that in the default case, with `MBEDTLS_X509_ALLOW_UNSUPPORTED_CRITICAL_EXTENSION` off we don't enforce critical certificate policy extensions that we can't interpret, and complete the verification of the chain with certificates that should fail.

In case 1. the user expects Mbed TLS to take care of certificate validation and to fail securely and we need to do so.

Reply:
I believe I fixed this with 8b0c3c91e63fee5d62c965f10cc35a760d5fea7f

Reply:
To allow users to dynamically check unknown policies in the verification callback, we need two things:
* The user needs a way to know that we've failed because of an unrecognized policy - 
I think `MBEDTLS_ERR_X509_FEATURE_UNAVAILABLE` is too generic for that. (However, we already have a similar issue with unknown, critical extensions [which is slightly different from what we're facing here, which is an unknown entry in a known, critical extension], because we return `MBEDTLS_ERR_X509_INVALID_EXTENSIONS + MBEDTLS_ERR_ASN1_UNEXPECTED_TAG`, which might also be returned while failing to parse a known extension).
* The user needs a way to inspect the policy that Mbed TLS didn't know about, which he can't at the moment because we don't include them in the CRT. Can we include _all_ policies in `crt->certificate_policies`?

Reply:
>The user needs a way to inspect the policy that Mbed TLS didn't know about, which he can't at the moment because we don't include them in the CRT. Can we include all policies in crt->certificate_policies?

They *are* included in `crt->certificate_policies`. What am I missing?